packages/contractkit/src/identity/offchain/schema-utils.ts

```diff
@@ -19,23 +42,39 @@ export class SingleSchema<T> {
   }
 }
 
+function foo(err: SchemaErrors) {
+  switch (err.errorType) {
+    case SchemaErrorTypes.OffchainError:
+      err
+      break
+
+    default:
+      break
+  }
+}
+
 export const readWithSchema = async <T>(
   wrapper: OffchainDataWrapper,
   type: t.Type<T>,
   account: Address,
   dataPath: string
-) => {
-  const data = await wrapper.readDataFrom(account, dataPath)
-  if (data === undefined) {
-    return
-  }
-  const asJson = JSON.parse(data)
-  const parseResult = type.decode(asJson)
-  if (isLeft(parseResult)) {
-    return undefined
+): Promise<Task<T, SchemaErrors>> => {
+  const resp = await wrapper.readDataFrom(account, dataPath)
+
+  if (isError(resp)) {
+    return Err(new OffchainError(resp.error))
   }
 
-  return parseResult.right
+  try {
+    const asJson = JSON.parse(resp.result)
+    const parseResult = type.decode(asJson)
+    if (isRight(parseResult)) {
+      return Ok(parseResult.right)
+    }
+    return Err(new RootError<SchemaErrorTypes.InvalidDataError>(SchemaErrorTypes.InvalidDataError))
+  } catch (error) {
+    return Err(new RootError<SchemaErrorTypes.InvalidDataError>(SchemaErrorTypes.InvalidDataError))
```

Comment:
I wonder if we can make things more concise, like removing the need for `new` and the duplication of the type and param.

Reply:
I agree that this is quite verbose, but I was unable to find a way to make it more concise. To my understanding:

- Constructing the Error with `new` is necessary so that the Error object will have the stacktrace
- The type annotation seemed necessary as the compiler was unable to deduce the proper type from a union type
- The param seemed necessary as the Error object itself needs the value for consumers to be able to pattern match on

I would love for someone with more typescript-fu to improve this

Reply:
Looks like extending the error interfaces with `Error` I was able to remove the type annotation!